File: completable-reactor-runtime/src/main/java/ru/fix/completable/reactor/runtime/validators/TerminalVertexExistValidator.java
package ru.fix.completable.reactor.runtime.validators;

import ru.fix.completable.reactor.api.ReactorGraphModel;
import ru.fix.completable.reactor.api.ReactorGraphModel.Transition;

/**
 * @author Kamil Asfandiyarov
 */
public class TerminalVertexExistValidator implements GraphValidator {

    @Override
    public void validateGraph(ReactorGraphModel graph) throws ValidationException {
        if (graph.getMergePoints().stream()
                .flatMap(mergePoint -> mergePoint.getTransitions().stream())
                .noneMatch(Transition::isComplete)) {

            throw new ValidationException("Graph should contain at least one terminal vertex." +
                    " Add complete step to at least one merge point.");
        }
    }
}

File: completable-reactor-runtime/src/main/kotlin/ru/fix/completable/reactor/runtime/validation/TerminalVertexExistValidator.kt
package ru.fix.completable.reactor.runtime.validation

import ru.fix.completable.reactor.runtime.internal.gl.GlReactorGraph

class TerminalVertexExistValidator: Validator {

    override fun validate(graph: GlReactorGraph<*>) {

        graph.vertices
                .asIterable()
                .any { it.tran }


        if (graph.getMergePoints().stream()
                        .flatMap({ mergePoint -> mergePoint.getTransitions().stream() })
                        .noneMatch(Predicate<ReactorGraphModel.Transition> { it.isComplete() })) {

            throw ValidationException("Graph should contain at least one terminal vertex." + " Add complete step to at least one merge point.")
        }
    }

    //TODO: fix injector
}